File: src/HTML_Logger/html_logger.cpp
#include <iostream>
#include <csignal>
#include <fstream>
#include <string>
#include <sstream>
#include <vector>
#include "html_logger.h"

using namespace std;

HTMLLogger::HTMLLogger(string output_file_path, string output_file_name, string project_type){
    this->output_file_path = output_file_path;
    this->output_file_name = output_file_name;
    this->project_type = project_type;
}

void HTMLLogger::create_file(){
    ifstream header_file;

    header_file.open(headerFilePath);
    if (!header_file){
        cout << "@HTMLLogger: Header file not found!" << endl;
        return;
    }

    output_file.open(output_file_path + output_file_name);

    // add opened header file into output file
    string line;
    while (std::getline(header_file, line)){
        output_file << line << '\n';
    }

    // append fuzz target heading
    if (project_type == "DJANGO"){
        output_file << "<p><b>Fuzz Target: </b>Django</p>";
    } else if (project_type == "COAP"){
        output_file << "<p><b>Fuzz Target: </b>CoAP</p>";
    } else if (project_type == "BLE"){
        output_file << "<p><b>Fuzz Target: </b>BLE</p>";
    }

    // TODO - display any other overall stats
}

// create table headings and update column size
void HTMLLogger::create_table_headings(string style, const vector<string> &column_names){
    column_num = column_names.size();

    output_file << R"( <table> )" << endl;
    output_file << R"( <tr style=")" << style << R"(;">)" << endl;
    for(int i = 0; i < column_num; i++){
        output_file << R"(<th>)" << column_names[i] << R"(</th>)" << endl;
    }
    output_file << R"(</tr>)" << endl;
}

// add_row without style
void HTMLLogger::add_row(const vector<string> &row){
    int row_size = row.size();

    if(row_size != column_num){
        cout << "@HTMLLogger: Invalid number of columns!" << endl;
        return;
    }

    output_file << "<tr>" << endl;
    for(int i = 0; i < row_size; i++){
        output_file << "<th>" << row[i] << "</th>" << endl;
    }
    output_file << "</tr>" << endl;
}

// add_row with style
void HTMLLogger::add_row(string style, const vector<string> &row){
    int row_size = row.size();

    if(row_size != column_num){
        cout << "@HTMLLogger: Invalid number of columns!";
        return;
    }

    output_file << "<tr>" << endl;
    for(int i = 0; i < row_size; i++){
        output_file << R"(<th style=")" << style << R"(;">)" << row[i] << R"(</th>)" << endl;
    }
    output_file << "</tr>" << endl;
}

void HTMLLogger::close_file(){
    ifstream footer_file;

    footer_file.open(footerFilePath);
    if (!footer_file){
        cout << "@HTMLLogger: Footer file not found!" << endl;
        return;
    }

    string line;
    while (getline(footer_file, line)){
        output_file << line << '\n';
    }
    output_file.close();
}

File: golang_src/HTML_Logger/html_logger.go
package html_logger

import (
	"errors"
	"fmt"
	"io/ioutil"
	"os"
)

type HTMLLogger struct {
	columnNum      int
	rowNum         int
	headerFilePath string
	footerFilePath string
	projectType    string
	outputFilePath string
	outputFileName string
	outputFile     *os.File
}

// Constructor
func NewHTMLLogger(outputFilePath string, outputFileName string, projectType string, outputFile *os.File) *HTMLLogger {
	return &HTMLLogger{
		rowNum:         0,
		columnNum:      0,
		headerFilePath: "./HTML_Logger/formats/header.html",
		footerFilePath: "./HTML_Logger/formats/footer.html",
		projectType:    projectType,
		outputFilePath: outputFilePath,
		outputFileName: outputFileName,
		outputFile:     outputFile,
	}
}

func (logger *HTMLLogger) CreateFile() error {
	headerFilePath := logger.headerFilePath

	fmt.Println("Header file path " + headerFilePath)

	headerFile, err := os.Open(headerFilePath)
	if err != nil {
		return errors.New("[ERROR] header file path failed to open")
	}

	outputFile, err := os.Create(logger.outputFilePath + logger.outputFileName)
	if err != nil {
		return errors.New("[ERROR] creating output file failed")
	}
	logger.outputFile = outputFile

	headerBytes, err := ioutil.ReadAll(headerFile)
	if err != nil {
		return errors.New("[ERROR] reading header file failed")
	}

	if _, err := outputFile.Write(headerBytes); err != nil {
		return errors.New("[ERROR] writing header content to output file")
	}

	switch logger.projectType {
	case "DJANGO":
		outputFile.WriteString("<p><b>Fuzz Target: </b>Django</p>")
	case "COAP":
		outputFile.WriteString("<p><b>Fuzz Target: </b>CoAP</p>")
	case "BLE":
		outputFile.WriteString("<p><b>Fuzz Target: </b>BLE</p>")
	}

	// TODO - display any other overall stats

	return nil
}

func (logger *HTMLLogger) AddText(style string, text string) {
	logger.outputFile.WriteString(fmt.Sprintf("<p style=\"%s;\">%s</p>\n", style, text))
}

func (logger *HTMLLogger) CreateTableHeadings(style string, columnNames []string) {
	logger.columnNum = len(columnNames)
	logger.outputFile.WriteString(" <table>\n")
	logger.outputFile.WriteString(fmt.Sprintf(" <tr style=\"%s;\">\n", style))
	logger.outputFile.WriteString("<th>Test no.</th>\n")

	for i := 0; i < logger.columnNum; i++ {
		logger.outputFile.WriteString(fmt.Sprintf("<th>%s</th>\n", columnNames[i]))
	}

	logger.outputFile.WriteString("</tr>\n")
}

func (logger *HTMLLogger) AddRow(row []string) {
	rowSize := len(row)

	logger.rowNum++

	if rowSize != logger.columnNum {
		fmt.Println("@HTMLLogger: Invalid number of columns!")
		return
	}

	logger.outputFile.WriteString("<tr>\n")
	logger.outputFile.WriteString(fmt.Sprintf("<th>%d</th>\n", logger.rowNum))

	for i := 0; i < rowSize; i++ {
		logger.outputFile.WriteString(fmt.Sprintf("<th>%s</th>\n", row[i]))
	}
	logger.outputFile.WriteString("</tr>\n")
}

func (logger *HTMLLogger) AddRowWithStyle(style string, row []string) {
	rowSize := len(row)

	logger.rowNum++

	if rowSize != logger.columnNum {
		fmt.Println("@HTMLLogger: Invalid number of columns!")
		return
	}

	logger.outputFile.WriteString("<tr>\n")
	logger.outputFile.WriteString(fmt.Sprintf(`<th style="%s">%d</th>`, style, logger.rowNum))

	for i := 0; i < rowSize; i++ {
		logger.outputFile.WriteString(fmt.Sprintf(`<th style="%s">%s</th>`, style, row[i]))
	}
	logger.outputFile.WriteString("</tr>")
}

// Close with footer (at least thats what i think it does - nic)
func (logger *HTMLLogger) CloseFile(footerFilePath string) error {
	footerFile, err := os.Open(footerFilePath)
	if err != nil {
		return fmt.Errorf("@HTMLLogger: Footer file not found! %v", err)
	}
	defer footerFile.Close()

	footerBytes, err := ioutil.ReadAll(footerFile)
	if err != nil {
		return fmt.Errorf("error reading footer file: %v", err)
	}
	if _, err := logger.outputFile.Write(footerBytes); err != nil {
		return fmt.Errorf("error writing footer content to output file: %v", err)
	}

	if err := logger.outputFile.Close(); err != nil {
		return fmt.Errorf("error closing output file: %v", err)
	}

	return nil
}
